utils/gcpcontrol/privateendpointscontroller.go

```diff
@@ -0,0 +1,365 @@
+	TargetServiceAttachment string
+	IPAddressName           string
+	ForwardingRuleName      string
+	DnsName                 string
+	ZoneName                string
+}
+
+type DnsChange struct {
+	Name    string
+	Type    string
+	Ttl     int64
+	Rrdatas []string
+}
+
+func (c *PrivateEndpointsController) GetInstanceWithoutZone(ctx context.Context, instanceID string) (*computepb.Instance, error) {
+	zonesClient, err := compute.NewZonesRESTClient(ctx)
+	if err != nil {
+		log.Fatalf("Failed to create zones client: %v", err)
+	}
+	defer zonesClient.Close()
+
+	zonesReq := &computepb.ListZonesRequest{
+		Project: c.ProjectID,
+	}
+	zonesIt := zonesClient.List(ctx, zonesReq)
+
+	var matchingZones []string
+	for {
+		zone, err := zonesIt.Next()
+		if err != nil {
+			break
+		}
+		if strings.HasPrefix(zone.GetName(), c.Region) {
+			matchingZones = append(matchingZones, zone.GetName())
+		}
+	}
+
+	// Try to find the instance in each zone
+	instancesClient, err := compute.NewInstancesRESTClient(ctx)
+	if err != nil {
+		log.Fatalf("Failed to create instances client: %v", err)
+	}
+	defer instancesClient.Close()
+
+	for _, zone := range matchingZones {
+		req := &computepb.GetInstanceRequest{
+			Project:  c.ProjectID,
+			Zone:     zone,
+			Instance: instanceID,
+		}
+
+		instance, err := instancesClient.Get(ctx, req)
+		if err == nil {
+			return instance, nil
+		}
+	}
+	return nil, errors.Wrap(err, "instance not found in any zone of configured region")
+}
+
+func (c *PrivateEndpointsController) GetInstanceUsingZone(ctx context.Context, instanceID, zone string) (*computepb.Instance, error) {
+	instancesClient, err := compute.NewInstancesRESTClient(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create instances client")
+	}
+	defer instancesClient.Close()
+
+	instance, err := instancesClient.Get(ctx, &computepb.GetInstanceRequest{
+		Zone:     zone,
+		Instance: instanceID,
+		Project:  c.ProjectID,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get instance details")
+	}
+
+	return instance, nil
+}
+
+func (c *PrivateEndpointsController) GetNetworkAndSubnet(ctx context.Context, instanceID, zone string) (*computepb.NetworkInterface, error) {
+	var instance *computepb.Instance
+	var err error
+	if zone == "" {
+		c.Logger.Info("zone not specified, looking for instance that matched instance-id in configured region")
+		instance, err = c.GetInstanceWithoutZone(ctx, instanceID)
+	} else {
+		c.Logger.Info("using specified zone to find instance")
+		instance, err = c.GetInstanceUsingZone(ctx, instanceID, zone)
+	}
+	if err != nil || instance == nil {
+		return nil, errors.Wrap(err, "failed to get instance details")
+	}
+	if len(instance.NetworkInterfaces) == 0 {
+		return nil, errors.New("instance has no network interfaces")
+	}
+	return instance.NetworkInterfaces[0], nil
+}
+
+func (c *PrivateEndpointsController) GetServiceAttachments(command string) (*GetServiceAttachmentsResult, error) {
```

Comment:
I guess this is still the only way to get the serivce attachments. This is good for now imo, but they should be adding an api to get these.

Reply:
Yeah, this is the only way to read service attachment. When I asked for some API, this is their reply - "We do have two service attachments but we need not expose their names."